File: backend/src/Models/Conversation.php
<?php

declare(strict_types=1);

namespace Quantis\AIPortfolioAssistant\Models;

/**
 * Represents a conversation with message history
 */
class Conversation
{
    private string $id;
    private ?string $userId;
    private array $messages = [];
    private \DateTimeImmutable $createdAt;
    private ?\DateTimeImmutable $updatedAt;
    private array $metadata;

    public function __construct(
        ?string $id = null,
        ?string $userId = null,
        array $metadata = []
    ) {
        $this->id = $id ?? uniqid('conv_');
        $this->userId = $userId;
        $this->createdAt = new \DateTimeImmutable();
        $this->updatedAt = null;
        $this->metadata = $metadata;
    }

    /**
     * Add a message to the conversation
     */
    public function addMessage(Message $message): self
    {
        $this->messages[] = $message;
        $this->updatedAt = new \DateTimeImmutable();
        return $this;
    }

    /**
     * Add a user message
     */
    public function addUserMessage(string $content): self
    {
        return $this->addMessage(Message::user($content));
    }

    /**
     * Add an assistant message
     */
    public function addAssistantMessage(string $content, array $metadata = []): self
    {
        return $this->addMessage(Message::assistant($content, $metadata));
    }

    /**
     * Get all messages
     */
    public function getMessages(): array
    {
        return $this->messages;
    }

    /**
     * Get message history as array for API
     */
    public function getHistory(): array
    {
        return array_map(fn(Message $m) => $m->toArray(), $this->messages);
    }

    /**
     * Get the last N messages
     */
    public function getLastMessages(int $count): array
    {
        return array_slice($this->messages, -$count);
    }

    /**
     * Get the last message
     */
    public function getLastMessage(): ?Message
    {
        return end($this->messages) ?: null;
    }

    /**
     * Clear all messages
     */
    public function clear(): self
    {
        $this->messages = [];
        $this->updatedAt = new \DateTimeImmutable();
        return $this;
    }

    /**
     * Get message count
     */
    public function getMessageCount(): int
    {
        return count($this->messages);
    }

    public function getId(): string
    {
        return $this->id;
    }

    public function getUserId(): ?string
    {
        return $this->userId;
    }

    public function getCreatedAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?\DateTimeImmutable
    {
        return $this->updatedAt;
    }

    public function getMetadata(): array
    {
        return $this->metadata;
    }

    public function setMetadata(array $metadata): self
    {
        $this->metadata = $metadata;
        return $this;
    }

    /**
     * Create from message history array
     */
    public static function fromHistory(array $history, ?string $userId = null): self
    {
        $conversation = new self(null, $userId);

        foreach ($history as $msg) {
            $conversation->addMessage(Message::fromArray($msg));
        }

        return $conversation;
    }

    /**
     * Export to array
     */
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'user_id' => $this->userId,
            'messages' => $this->getHistory(),
            'message_count' => $this->getMessageCount(),
            'created_at' => $this->createdAt->format('c'),
            'updated_at' => $this->updatedAt?->format('c'),
            'metadata' => $this->metadata,
        ];
    }
}

File: backend_python/app/models/conversation.py
"""Conversation model."""
from datetime import datetime, timezone
from app.support.phpcompat import php_uniqid
from .message import Message


class Conversation:
    """Represents a conversation with message history."""

    def __init__(
        self,
        id: str = None,
        user_id: str = None,
        metadata: dict = None,
    ):
        """Initialize a conversation."""
        self.id = id or php_uniqid('conv_')
        self.user_id = user_id
        self.messages = []
        self.created_at = datetime.now(timezone.utc)
        self.updated_at = None
        self.metadata = metadata or {}

    def addMessage(self, message: Message) -> 'Conversation':
        """Add a message to the conversation."""
        self.messages.append(message)
        self.updated_at = datetime.now(timezone.utc)
        return self

    def addUserMessage(self, content: str) -> 'Conversation':
        """Add a user message."""
        return self.addMessage(Message.user(content))

    def addAssistantMessage(self, content: str, metadata: dict = None) -> 'Conversation':
        """Add an assistant message."""
        return self.addMessage(Message.assistant(content, metadata))

    def getMessages(self) -> list:
        """Get all messages."""
        return self.messages

    def getHistory(self) -> list:
        """Get message history as array for API."""
        return [m.toArray() for m in self.messages]

    def getLastMessages(self, count: int) -> list:
        """
        Get the last N messages using PHP array_slice($messages, -$count) semantics:
        - count 3 → last three
        - count 0 → the whole list
        - count -2 → everything from index 2
        """
        return self.messages[-count:] if count else self.messages

    def getLastMessage(self) -> Message | None:
        """Get the last message."""
        return self.messages[-1] if self.messages else None

    def clear(self) -> 'Conversation':
        """Clear all messages."""
        self.messages = []
        self.updated_at = datetime.now(timezone.utc)
        return self

    def getMessageCount(self) -> int:
        """Get message count."""
        return len(self.messages)

    def getId(self) -> str:
        """Get the conversation ID."""
        return self.id

    def getUserId(self) -> str | None:
        """Get the user ID."""
        return self.user_id

    def getCreatedAt(self) -> datetime:
        """Get the created timestamp."""
        return self.created_at

    def getUpdatedAt(self) -> datetime | None:
        """Get the updated timestamp."""
        return self.updated_at

    def getMetadata(self) -> dict:
        """Get the conversation metadata."""
        return self.metadata

    def setMetadata(self, metadata: dict) -> 'Conversation':
        """Set the conversation metadata."""
        self.metadata = metadata
        return self

    @staticmethod
    def fromHistory(history: list, user_id: str = None) -> 'Conversation':
        """Create from message history array."""
        conversation = Conversation(None, user_id)

        for msg in history:
            conversation.addMessage(Message.fromArray(msg))

        return conversation

    def toArray(self) -> dict:
        """Export to array."""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'messages': self.getHistory(),
            'message_count': self.getMessageCount(),
            'created_at': self.created_at.isoformat(timespec='seconds'),
            'updated_at': self.updated_at.isoformat(timespec='seconds') if self.updated_at else None,
            'metadata': self.metadata,
        }
